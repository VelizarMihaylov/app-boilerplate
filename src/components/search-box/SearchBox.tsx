import React, { useState } from 'react'
import { useDispatch } from 'react-redux'

import { MagnifyingGlassIcon } from './icon'
import styled from '@emotion/styled'
import { spinner, gridUnits, colours } from 'mixins'

import { useQuery } from '@apollo/react-hooks'
import { GET_CITIES } from 'graph-ql'

const SearchBoxContainer = styled.div`
  width: 100%;
  position: relative;
  height: ${gridUnits(2.5)}rem;
`

const SearchBoxForm = styled.form`
  position: absolute;
  z-index: 999;
  margin-left: 50%;
  transform: translateX(-50%);
  width: ${gridUnits(20)}rem;
  height: ${gridUnits(2.5)}rem;
  border: ${gridUnits(0.2)}rem solid grey;
  border-radius: ${gridUnits(0.75)}rem;
  background: ${colours.white};
`

const SearchBoxInput = styled.input`
  position: absolute;
  top: 0;
  left: ${gridUnits(3)}rem;
  right: 0;
  font-size: ${gridUnits(1)}rem;
  line-height: ${gridUnits(2.5)}rem;
  background: none;
  width: ${gridUnits(16.5)}rem;
  height: ${gridUnits(2.5)}rem;
  border: none;
  appearance: none;
  outline: none;

  &::placeholder {
    color: ${colours.black};
  }
`

const SearchBoxDatalist = styled.ul`
  position: absolute;
  top: ${gridUnits(2)}rem;
  width: ${gridUnits(18.5)}rem;
  margin-left: 50%;
  transform: translateX(-50%);
  max-height: ${gridUnits(10)}rem;
  overflow-y: scroll;
  list-style: none;
  background: white;
  border: ${gridUnits(0.2)}rem solid grey;
  border-top: none;
  border-radius: 0 0 ${gridUnits(0.5)}rem ${gridUnits(0.5)}rem;
  z-index: 998;

  & li {
    width: 100%;
    padding: ${gridUnits(1.5)}rem 0 ${gridUnits(0.75)}rem 0;
    text-align: center;
    cursor: pointer;
  }

  & li:hover {
    background: ${colours.grey};
  }
`

const StyledMagnifyingGlassIcon = styled(MagnifyingGlassIcon)`
  position: absolute;
  top: ${gridUnits(0.5)}rem;
  left: ${gridUnits(0.75)}rem;
  height: ${gridUnits(1.5)}rem;
  width: ${gridUnits(1.5)}rem;
  padding: 0;
  margin: 0;
  border: none;
  background: none;
  fill: ${colours.grey};
  outline: none !important;
`
const SearchBoxIconSpinner = styled.div`
  ${spinner({
    border: `2px solid ${colours.grey}`,
    topBorder: `2px solid ${colours.white}`,
    width: 1.5,
    height: 1.5
  })};
  position: absolute;
  top: ${gridUnits(0.5)}rem;
  left: ${gridUnits(0.75)}rem;
  padding: 0;
  margin: 0;
  background: none;
  outline: none !important;
`

type QueryDataResponse = {
  cities: {
    name: string
  }[]
}
export const SearchBox: React.FC = (): React.ReactElement => {
  const { data, loading, error } = useQuery<QueryDataResponse>(GET_CITIES)
  const dispatch = useDispatch()
  const [dataList, setDataList] = useState<QueryDataResponse['cities']>([])
  const [selection, setSelection] = useState('')
  const [showList, setShowList] = useState(false)
  if (error) return <h1>Oops something went wrong!</h1>
  const { cities } = data || { cities: [] }
  return (
    <SearchBoxContainer>
      <SearchBoxForm data-puppet="search-box-form">
        {loading ? <SearchBoxIconSpinner /> : <StyledMagnifyingGlassIcon />}
        <SearchBoxInput
          data-puppet="search-box-input"
          disabled={loading}
          className="SearchBox--input"
          autoComplete="off"
          name="cities"
          placeholder={loading ? undefined : 'Enter city name...'}
          onChange={(event: React.ChangeEvent<HTMLInputElement>): void => {
            setDataList(
              cities.filter(({ name }) =>
                name.match(RegExp(event.target.value, 'i'))
              )
            )
            setSelection(event.target.value)
          }}
          onBlur={(): void => setShowList(false)}
          onFocus={(): void => {
            if (!dataList.length) setDataList(cities)
            setShowList(true)
          }}
          value={selection}
        />
      </SearchBoxForm>
      <SearchBoxDatalist
        style={{
          display: showList ? 'block' : 'none'
        }}
      >
        {dataList.map(({ name }) => (
          <li
            data-puppet="search-box-list-element"
            key={name}
            onMouseDown={(): void => {
              setSelection(name)
              dispatch({ type: 'ADD_CITY', payload: name })
            }}
          >
            {name}
          </li>
        ))}
      </SearchBoxDatalist>
    </SearchBoxContainer>
  )
}
