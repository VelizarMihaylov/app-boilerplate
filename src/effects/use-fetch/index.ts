import { useReducer, useEffect, useState, Reducer } from 'react'

type Action<P> =
  | { type: 'FETCH_INIT' }
  | {
      type: 'FETCH_SUCCESS'
      payload: P
    }
  | {
      type: 'FETCH_FAILURE'
    }

type State<P> = {
  loading: boolean
  data: null | P
  error: boolean
}

const dataFetchReducer = <P>(state: State<P>, action: Action<P>): State<P> => {
  switch (action.type) {
    case 'FETCH_INIT':
      return {
        ...state,
        loading: true
      }
    case 'FETCH_SUCCESS':
      return {
        ...state,
        loading: false,
        error: false,
        data: action.payload
      }
    case 'FETCH_FAILURE':
      return {
        ...state,
        loading: false,
        error: true
      }
    default:
      throw new Error(
        `Unknown action type received 
        for dataFetchReducer.
        Please make sure you are passing
        one of the following actions:
          * FETCH_INIT
          * FETCH_SUCCESS
          * FETCH_FAILURE
          :`
      )
  }
}

type useFetchResult<P> = {
  loading: boolean
  data: null | P
  error: boolean
  reload: () => void
  retries: number
  fetchLazy: (url: string) => void
}

export const useFetch = <P>(url: undefined | string): useFetchResult<P> => {
  const initialState = {
    loading: false,
    data: null,
    error: false
  }
  const [state, dispatch] = useReducer<Reducer<State<P>, Action<P>>>(
    dataFetchReducer,
    initialState
  )

  const [reload, setReload] = useState<number>(0)
  const [link, setLink] = useState<undefined | string>(url)

  useEffect(() => {
    if (link) {
      dispatch({ type: 'FETCH_INIT' })
      fetch(link)
        .then(response => {
          return response.json()
        })
        .then((data: P) => {
          dispatch({ type: 'FETCH_SUCCESS', payload: data })
        })
        .catch(error => {
          if (error.name === 'AbortError') {
            console.log('User aborted the fetch')
          } else {
            console.log('ERROR', error.name)
            dispatch({ type: 'FETCH_FAILURE' })
          }
        })
    }
  }, [link, reload])

  return {
    ...state,
    reload: (): void => {
      setReload(reload + 1)
    },
    retries: reload,
    fetchLazy: (url: string): void => {
      setLink(url)
    }
  }
}
