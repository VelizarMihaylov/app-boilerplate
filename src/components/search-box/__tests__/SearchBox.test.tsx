import React from 'react'
import { SearchBox } from '../SearchBox'
import { create, act } from 'react-test-renderer'

import { useFetch } from 'effects/use-fetch'
jest.mock('effects/use-fetch')

const mockedUseFetch = useFetch as jest.Mock<unknown>

import { useDispatch } from 'react-redux'
jest.mock('react-redux')

const mockedUseDispatch = useDispatch as jest.Mock<unknown>

describe('SearchBox', () => {
  afterEach(() => jest.resetAllMocks())
  const data = {
    results: [
      {
        country: 'GB',
        name: 'London',
        city: 'London',
        count: 158120,
        locations: 3
      },
      {
        country: 'GB',
        name: 'Manchester',
        city: 'Manchester',
        count: 48890,
        locations: 1
      },
      {
        country: 'GB',
        name: 'Leeds',
        city: 'Leeds',
        count: 51077,
        locations: 1
      }
    ]
  }
  it('should handle loading state', () => {
    mockedUseFetch.mockImplementation(() => ({ loading: true }))
    mockedUseDispatch.mockImplementation(() => jest.fn())
    const SearchBoxRender = create(
      <SearchBox url="https://test-api.com" />
    ).toJSON()
    expect(SearchBoxRender).toMatchSnapshot()
  })
  it('should handle error state', () => {
    mockedUseFetch.mockImplementation(() => ({ error: true }))
    mockedUseDispatch.mockImplementation(() => jest.fn())
    const SearchBoxRender = create(
      <SearchBox url="https://test-api.com" />
    ).toJSON()
    expect(SearchBoxRender).toMatchSnapshot()
  })
  it('should list all cities options when onFocus event is triggered and data is fetched', () => {
    mockedUseFetch.mockImplementation(() => ({
      loading: false,
      error: false,
      data
    }))
    mockedUseDispatch.mockImplementation(() => jest.fn())
    const SearchBoxRender = create(<SearchBox url="https://test-api.com" />)

    const input = SearchBoxRender.root.find(element => element.type === 'input')

    act(() => {
      input.props.onFocus()
    })

    expect(
      SearchBoxRender.root.find(element => element.type === 'ul').children
        .length
    ).toBe(3)
  })

  it('should dispatch ADD_CITY action when the city list is empty and onFocus action is triggered', () => {
    mockedUseFetch.mockImplementation(() => ({
      loading: false,
      error: false,
      data
    }))
    const dispatch = jest.fn()
    mockedUseDispatch.mockImplementation(() => dispatch)
    const SearchBoxRender = create(<SearchBox url="https://test-api.com" />)

    const input = SearchBoxRender.root.find(element => element.type === 'input')

    act(() => {
      input.props.onFocus()
    })

    act(() => {
      SearchBoxRender.root
        .find(element => element.type === 'ul')
        .props.children[0].props.onMouseDown()
    })

    expect(dispatch).toHaveBeenCalledWith({
      type: 'ADD_CITY',
      payload: 'London'
    })
  })
})
