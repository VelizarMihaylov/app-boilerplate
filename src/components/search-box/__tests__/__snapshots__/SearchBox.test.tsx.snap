// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox should handle error state 1`] = `
<h1>
  Oops something went wrong!
</h1>
`;

exports[`SearchBox should handle loading state 1`] = `
<div
  className="css-7410gz"
>
  <form
    className="css-aq80jo"
  >
    <div
      className="css-1g181gq"
    />
    <input
      autoComplete="off"
      className="SearchBox--input css-1tsdkxm"
      disabled={true}
      name="cities"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value=""
    />
  </form>
  <ul
    className="css-bdlxgg"
    style={
      Object {
        "display": "none",
      }
    }
  />
</div>
`;
